cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -19,152 +22,211 @@ class UpgradeCommand < Kontena::Command
 
     option '--force', :flag, 'Force upgrade'
     option '--skip-dependencies', :flag, "Do not install any stack dependencies"
-    option '--dry-run', :flag, "Do not perform any uploading", hidden: true
+    option '--dry-run', :flag, "Simulate upgrade"
 
     requires_current_master
     requires_current_master_token
 
-    def normalize_local_data(stack_data, parent_name)
-      return nil if stack_data.nil? || stack_data.empty?
-
-      depends = stack_data.delete('depends') || []
-      normalized_data = {
-        parent_name => stack_data.merge(
-          :loader => loader_class.for(stack_data['stack'])
-        )
-      }
+    # @return [Kontena::Cli::Stacks::ChangeResolver]
+    def execute
+      set_env_variables(stack_name, current_grid)
 
-      depends.each do |stack|
-        key = "#{parent_name}-#{stack['name']}"
-        normalized_data.merge!(normalize_local_data(stack.merge('parent_name' => parent_name), key))
+      old_data = spinner "Reading stack #{pastel.cyan(stack_name)} from master" do
+        gather_master_data(stack_name)
       end
-      normalized_data
-    end
 
-    def normalize_master_data(stack_name, raise_not_found = false)
-      begin
-        data = fetch_master_data(stack_name)
-      rescue Kontena::Errors::StandardError => ex
-        return nil if ex.status == 404 && !raise_not_found
-        raise ex
+      new_data = spinner "Parsing #{pastel.cyan(source)}" do
+        loader.flat_dependencies(
+          stack_name,
+          variables: values_from_options
+        )
       end
-      depends = data.delete('children') || []
 
-      normalized_data = { stack_name => data }
+      changes = process_data(old_data, new_data)
 
-      return normalized_data if skip_dependencies?
+      display_report(changes)
 
-      depends.each do |stack|
-        normalized_data.merge!(normalize_master_data(stack['name']))
-      end
-      normalized_data
+      return if dry_run?
+
+      get_confirmation(changes)
+
+      run_removes(changes.removed_stacks)
+
+      deployable_stacks = []
+      deployable_stacks.concat run_installs(changes)
+      deployable_stacks.concat run_upgrades(changes)
+
+      run_deploys(deployable_stacks) if deploy?
```

Comment:
I suspect that the new ordering of the remove -> install -> upgrade -> deploy operations here isn't as good as the current order of operations. Running the installs and upgrades separately before running all of the deploys as a separate step at the end means that upgrades of each stack might not actually see the installed/upgraded versions of the earlier stacks.

I think the best way would be to install/upgrade + deploy each child stack in the same order as defined in the YAML, before upgrading the parent stack. Any old stacks should then probably be removed at the end.

Reply:
Could be. I separated the deploy because it seemed like a good idea to install everything properly first and then trigger the deploys, but I'm not any sort of wizard with the server/agent stuff.

Reply:
Running the removes last certainly sounds like a good idea.